ethcore/builtin/src/lib.rs

```diff
@@ -211,89 +238,187 @@ impl Builtin {
+		};
+
+		let pricer = match b.pricing {
+			ethjson::spec::builtin::Pricing::Single(pricer) => {
+				let mut pricers = BTreeMap::new();
+
+				let (pricing, eip1108_pricing) = into_pricing(pricer);
+
+				if b.activate_at.is_none() {
+					warn!(target: "builtin",
+						"Builtin contract: {} is missing which block to activate it on, failing back to default: 0",
+						b.name
+					);
+				}
+
+				pricers.insert(b.activate_at.map_or(0, Into::into), pricing);
+
+				if let (Some(activate), Some(p)) = (b.eip1108_transition, eip1108_pricing) {
+					pricers.insert(activate.into(), p);
+					warn!(target: "builtin",
+						"Builtin contract: {} enabled with eip1108_transition which is deprecated. \
+						Use builtin contract with multiple activations instead in your chain specification",
+						b.name
+					);
+				}
+				pricers
 			}
-			ethjson::spec::Pricing::Modexp(exp) => {
-				Box::new(ModexpPricer {
-					divisor: if exp.divisor == 0 {
-						warn!(target: "builtin", "Zero modexp divisor specified. Falling back to default.");
-						10
-					} else {
-						exp.divisor
+
+			ethjson::spec::builtin::Pricing::Multi(pricer) => {
+				assert!(!pricer.is_empty(), "Chainspec with builtin contract with multiple activations was empty");
+
+				let mut pricers = BTreeMap::new();
+				for p in pricer {
+					let (pricing, eip1108_pricing) = into_pricing(p.price);
+
+					if eip1108_pricing.is_some() {
+						warn!(target: "builtin",
+							"Builtin contract: {} is with configured multiple activations but contains \
+							eip1108_transition which is not supported. The contract is activated on block: {:?}",
+							b.name, p.activate_at
+						);
 					}
-				})
-			}
-			ethjson::spec::Pricing::AltBn128Pairing(pricer) => {
-				Box::new(AltBn128PairingPricer {
-					price: AltBn128PairingPrice {
-						base: pricer.base,
-						pair: pricer.pair,
-					},
-					eip1108_transition_at: b.eip1108_transition.map_or(u64::max_value(), Into::into),
-					eip1108_transition_price: AltBn128PairingPrice {
-						base: pricer.eip1108_transition_base,
-						pair: pricer.eip1108_transition_pair,
-					},
-				})
-			}
-			ethjson::spec::Pricing::AltBn128ConstOperations(pricer) => {
-				Box::new(AltBn128ConstOperations {
-						price: pricer.price,
-						eip1108_transition_price: pricer.eip1108_transition_price,
-						eip1108_transition_at: b.eip1108_transition.map_or(u64::max_value(), Into::into)
-				})
+
+					pricers.insert(p.activate_at.into(), pricing);
+				}
+				pricers
 			}
 		};
+		Self { pricer, native }
+	}
+}
 
-		Builtin {
-			pricer,
-			native: ethereum_builtin(&b.name),
-			activate_at: b.activate_at.map_or(0, Into::into),
+// Convert `JSON pricing` into Pricing
+//
+// Because `AltBn128XX` has two different pricings for `eip1108 transition` both need
+// to be taken into account.
+//
+// Returns `(Pricing, Some(Pricing))` if eip1108_transition is enabled with an activation
+// otherwise it returns `(Pricing, None)`
```

Comment:
Yes, I agree. 
It would be good if we could remove the mess from #11008 but it might annoy people that recently updated their chain spec except maybe the type `Const/Fixed price` instead of Linear. If we decide to remove the hack we must add something in the changelog. 

Lesson: don't write hacks that are supposed to be backward compatible

Reply:
I guess it depends a bit on whether we want this backported or not. I think we do want this backported (but unsure tbh) and in that case we should avoid breaking peoples setups. We could also opt to not backport it and have this go in to the 2.7 release right after the Istanbul HF, in which case it's fine to break things.

Reply:
I think it will become too hard to maintain and a lot of ugliness. Let's keep the old format but deprecate the `eip1108 params` and don't backport?

Reply:
> Let's keep the old format but deprecate the eip1108 params and don't backport?

Works for me. Ok by you @ordian ?

Reply:
> And by deprecate you don't mean remove, 

Sorry, I meant remove but I reverted it now, 

> so it will add complexity to the code

https://github.com/paritytech/parity-ethereum/pull/11039/commits/d58ef6198663e17423f04057d4e2652e85be789b is the complexity it introduces

> I would rather go for "annoy people that recently updated their chain spec" approach to break compatibility (but show them how to upgrade in the release notes), but only do it once with a future-proof API.

Yes, I agree